javalib/src/main/scala/java/util/AbstractList.scala

```diff
@@ -1,3 +1,293 @@
+      case (hashCode, elem) => 31 * hashCode + elem.hashCode
+    }
+  }
+
+  def indexOf(o: Any): Int =
+    iterator().indexWhere(_ === o)
+
+  def iterator(): Iterator[E] =
+    listIterator()
+
+  def lastIndexOf(o: Any): Int = {
+    @tailrec
+    def findIndex(iter: ListIterator[E]): Int = {
+      if (!iter.hasPrevious) -1
+      else if (iter.previous() === o) iter.nextIndex
+      else findIndex(iter)
+    }
+    findIndex(listIterator(size))
+  }
+
+  def listIterator(): ListIterator[E] =
+    listIterator(0)
+
+  def listIterator(index: Int): ListIterator[E] = {
+    checkIndexOnBounds(index)
+    new RandomAccessListIterator(index, 0, size)
+  }
+
+  def remove(index: Int): E =
+    throw new UnsupportedOperationException
+
+  protected def removeRange(fromIndex: Int, toIndex: Int): Unit = {
+    @tailrec
+    def removeNext(index: Int, iter: ListIterator[E]): Unit = {
+      if (iter.hasNext && index < toIndex) {
+        println(iter.next())
+        iter.remove()
+        removeNext(index + 1, iter)
+      }
+    }
+    removeNext(fromIndex, listIterator(fromIndex))
+  }
+
+  def set(index: Int, element: E): E =
+    throw new UnsupportedOperationException
+
+  def subList(fromIndex: Int, toIndex: Int): List[E] = {
+    if (fromIndex < 0 || toIndex > size)
+      throw new IndexOutOfBoundsException
+    else if (fromIndex > toIndex)
+      throw new IllegalArgumentException
+
+    self match {
+      case _: RandomAccess =>
+        new AbstractListView(fromIndex, toIndex) with RandomAccess {
+          override def listIterator(index: Int): java.util.ListIterator[E] = {
+            checkIndexOnBoundsOfView(index)
+            // Iterator that accesses the original list directly
+            new RandomAccessListIterator(fromIndex + index, fromIndex, toIndex)
+          }
+        }
+      case _ =>
+        new AbstractListView(fromIndex, toIndex) {
+          override def listIterator(index: Int): java.util.ListIterator[E] = {
+            checkIndexOnBoundsOfView(index)
+            // Iterator that accesses the original list using it's iterator
+            new BackedUpListIterator(index, fromIndex)
+          }
+        }
+    }
+  }
+
+  private def checkIndexOnBounds(index: Int): Unit = {
+    if (index < 0 || index > size)
+      throw new IndexOutOfBoundsException(index.toString)
+  }
+
+  private abstract class AbstractListView(fromIndex: Int,
+                                          private var toIndex: Int) extends AbstractList[E] {
+
+    override def add(e: E): Boolean = {
+      add(size, e)
+      toIndex += 1
+      true
+    }
+
+    override def add(index: Int, e: E): Unit = {
+      checkIndexOnBoundsOfView(index)
+      self.add(fromIndex + index, e)
+    }
+
+    override def addAll(index: Int, c: Collection[_ <: E]): Boolean = {
+      checkIndexOnBoundsOfView(index)
+      val success = self.addAll(fromIndex + index, c)
+      if (success)
+        toIndex += c.size
+      success
```

Comment:
But your code only avoids calling `c.size` if it is in fact 0. But if `c.size` is 0, then traversing the collection is O(1) as well.